### ISSUE TITLE
Publish updated WD

### ISSUE THREAD

**Issue by mmccool** [w3c/wot-discovery#154]
Please leave comments on the proposed publication of an updated WD here.  Copy of email calling for resolution:
https://lists.w3.org/Archives/Public/public-wot-wg/2021Apr/0000.html

**Comment by wiresio** [comment]
Two points - summarizing what I tried to point out in the calls and several other issues:
- Although it is a good starting point comprising helpful hints for developers, the current spec. is too much centering on http (e.g. http 1.1 chunked transfer encoding https://github.com/w3c/wot-discovery/pull/145, http header based pagination https://github.com/w3c/wot-discovery/pull/153) while there is demand for support of other protocols like CoAP and MQTT https://github.com/w3c/wot-discovery/issues/134.
Proposal: Mark http specific parts as informative and point out that spec. aims at supporting other protocols as well. For the reference implemations it would relieve us from the burden of testing protocol details instead of focusing on testing (core) TDD features.
- Lack of support for sending server side information along with data (i.e. TDs) in the listing API and potentially (not yet decided) in the search APIs: This prevents a TDD server from sending e.g. pagination information in the body. While I understand that the very puristic and clean solution requires a JSON-LD formatted response and as a consequence forbids embedding JSON-LD in JSON, I'm wondering whether it is sufficient for all use cases. In fact the only use case I see for the puristic approach is "stream TDs from one directory to another". I'd like to have a possibility for the server to provide additional information to the clients (not strictly limited to pagination information).
Proposal: Define a JSON "envelope format" like in https://github.com/w3c/wot-discovery/pull/124 or https://www.w3.org/community/hydra/wiki/Pagination#PartialCollection in addition to the puristic JSON-LD format for both listing and search APIs and avoid to mandate too much protocol specifics (see point 1).

**Comment by wiresio** [comment]
Some other points:

- 6.1: I propose to add one sentence which makes clear that self-description using protocols other than http is still to be defined such that we do not create the impression of excluding other protocols.
- 6.2.1: Yet another idea for allowing server-side information: What about adding a JSON-LD container tdd:ServerInfo similar to tdd:LinkDescription?
- 6.2.2: "The HTTP API MUST be exposed over HTTPS (HTTP Over TLS)." -> Should also made clear that http is one protocol choice, potentially the preferred one. For other protocols appropriate TLS has to be applied.
- Example 3: Not normative as stated in §2, right?
- 6.2.2.1: "Registration API is a RESTful HTTP API" -> again just http
- Issue 48: Should be left up to the implementation.
- 6.2.2.4: Is this the place for my pagination dreams to come true ;-) ?

**Comment by wiresio** [comment]
How JSON-LD collections should be handled:  
https://w3c.github.io/json-ld-bp/#serializing-large-collections -> https://www.hydra-cg.com/spec/latest/core/#collections

**Comment by mmccool** [comment]
Summary of discussion of second "envelope" point:
0. Some devs don't want to use header-based chunking, and it is also dependent on HTTP
1. Mixing content types is incorrect, so putting JSON inside JSON-LD is not "legal"
2. We could use an envelope which is JSON-LD (however, see previous discussions; @farshid, can you summarize)
3. We could use a JSON envelope which uses pointers, but requires additional fetches
Action:
- Create a concrete PR that presents an option for envelope-based pagination
- Defer deciding whether it is optional or mandatory on directory

**Comment by mmccool** [comment]
Also, regarding the first point mentioned, will add an explanation (issue link) about why we are focused on HTTP.

**Comment by mmccool** [comment]
As a further comment, it has often been discussed that the reason we don't necessarily "need" a JSON envelope-based paging scheme (e.g. not using headers, and so protocol independent, etc.) is that the equivalent effect can be obtained through a specific set of JSONPath queries that select ranges of TDs.  This would also be protocol independent, assuming JSON Path queries will work with other protocols.

So another option here is to make that a bit more "official" by adding a subsection to the pagination section that elaborates upon it, provides examples, recommended query patterns, etc.  This subsection could be marked as informative since it does not itself add any new features.

**Comment by wiresio** [comment]
I started a discussion with the Hydra CG: https://github.com/HydraCG/Specifications/issues/236   
Please feel free to comment there.

**Comment by farshidtz** [comment]
I just want to clarify that resource-based pagination is NOT an alternative to HTTP chunked transfer encoding. Note that I call this resource-based (and not header-based) because the headers in the current spec are just informative; we use query arguments to navigate. Even with resource-based pagination, we still need to do streaming (using protocol-specific mechanisms) to handle large TDs. **Streaming is not mandatory.** Strictly speaking, the alternative to pagination is "no pagination". 

The payload data model is currently array of TDs for the HTTP response body (listing API). An alternative payload format is the envelope with array of TDs and pagination info inside. Sending data in streams applies regardless of the payload format. Saying that we have a specific HTTP/1.1 format is not correct. We have two **recommendations** for streaming/chunks in **HTTP**: one for HTTP/1.1 (chunked transfer encoding) and another for HTTP/2 (frames). They are from HTTP standards. **Neither of the two are mandatory.**

I suggest creating an issue to first discuss the shortcomings of the current data model, in particular to address what hinders use cases from efficiently retrieving the TDs from the directory.

**Comment by mmccool** [comment]
Done; WD2 now published.  See https://www.w3.org/TR/wot-discovery/ published June 2.
